File: python/paypal_agent_toolkit/langchain/toolkit.py
"""Toolkit for LangChain."""

from typing import Dict, List, Any, Optional, Union

from ..configuration import Context
from .tool import create_langchain_tools


class PayPalToolkit:
    """Toolkit for PayPal API integration with LangChain."""

    _tools: List[Dict[str, Any]]

    def __init__(
        self,
        client_id: str,
        client_secret: str,
        allowed_tools: Optional[List[str]] = None,
        context: Optional[Context] = None,
    ):
        """Initialize the PayPal toolkit.
        
        Args:
            client_id: The PayPal client ID.
            client_secret: The PayPal client secret.
            allowed_tools: Optional list of allowed tool names.
            context: Optional context for the API.
        """
        self._tools = create_langchain_tools(client_id, client_secret, allowed_tools, context)

    @property
    def tools(self) -> List[Dict[str, Any]]:
        """Get the LangChain tools for this toolkit."""
        return self._tools

    def get_tools(self) -> List[Any]:
        """Get the LangChain tools for this toolkit.
        
        Returns:
            A list of LangChain tool objects.
        """
        try:
            # Try to import LangChain
            from langchain.tools import Tool
            
            # Convert our tool definitions to LangChain Tool objects
            lc_tools = []
            for tool in self._tools:
                lc_tool = Tool(
                    name=tool["name"],
                    description=tool["description"],
                    func=tool["func"],
                    args_schema=tool["args_schema"],
                    return_direct=tool["return_direct"],
                )
                lc_tools.append(lc_tool)
            
            return lc_tools
        except ImportError:
            raise ImportError(
                "LangChain is required to use this toolkit. "
                "Please install it with `pip install langchain`."
            )

File: typescript/src/langchain/toolkit.ts
/**
 * PayPal Agent LangChain Toolkit in TypeScript
 */

import {BaseToolkit} from "@langchain/core/tools";
import PayPalAPI from "../shared/api";
import PayPalClient from "../shared/client";
import { Configuration, isToolAllowed } from "../shared/configuration";
import tools from "../shared/tools";
import PayPalTool from "./tool";

const SOURCE = "LANGCHAIN";

class PayPalLangChainToolkit implements BaseToolkit {
    readonly client: PayPalClient;
    private _paypal: PayPalAPI;
    tools: PayPalTool[];

    constructor({ clientId, clientSecret, configuration, }: {
        clientId: string,
        clientSecret: string,
        configuration: Configuration;
    }) {
        const context = configuration.context || {};
        this.client = new PayPalClient({ clientId: clientId, clientSecret: clientSecret, context: { ...context, source: SOURCE }});
        const filteredTools = tools(context).filter((tool) =>
            isToolAllowed(tool, configuration)
        ); 
        this._paypal = new PayPalAPI(this.client, configuration.context);
        this.tools = filteredTools.map(
            (tool) =>
                new PayPalTool(
                    this._paypal, 
                    tool.method,
                    tool.name,
                    tool.description,
                    tool.parameters,
                )
        );
    }

    getTools(): PayPalTool[] {
        return this.tools;
    }
}

export default PayPalLangChainToolkit;